File: App/Components/RingSticker.swift
import SwiftUI

// RingSticker — her flower rides the cycle ring like a bead on a bangle. The
// hitbox is the flower itself (the rest of the ring keeps its scrub gesture).
// Dragging along the ring band keeps it magnetically locked to the ring —
// always concentric, never offset; fling it and it keeps spinning like a
// fidget spinner, shedding confetti while it's fast. Pull it clearly off the
// band and it PLUCKS free: drop it inside the ring or off to the side and it
// rests there (persisted); carry it back to the band and it snaps on again.
// Reduce Motion: drag still works, no inertia, no trail.
struct RingSticker: View {
    @Environment(Theme.self) private var theme
    @Environment(RewardsStore.self) private var rewards
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    let radius: CGFloat
    /// Fraction of the ring (from 12 o'clock, clockwise) covered by the period
    /// arc — landing the flower there pays a small bonus.
    var periodFraction: Double = 0

    @State private var angle: Double = -0.9
    @State private var freePos: CGPoint = .zero    // center-origin, when plucked
    @State private var onRing = true
    @State private var plucked = false             // mid-drag, off the band
    @State private var velocity: Double = 0        // radians / frame
    @State private var dragging = false
    @State private var spinning = false
    @State private var lastDragAngle: Double? = nil
    @State private var trail: [TrailBit] = []
    @State private var settleBurst = 0
    @State private var spins = 0
    @State private var showTease = false
    @State private var showBonus = false

    /// Frame margin: wide enough that a free rest can clear the snap band on
    /// every axis (margin/2 − 14 > snapBand), so "beside the ring" is stable.
    private var diameter: CGFloat { radius * 2 + 96 }
    /// How far off the ring counts as "still on the band" (magnetic snap zone).
    private let snapBand: CGFloat = 26

    /// Where the flower sits right now, center-origin.
    private var pos: CGPoint {
        if plucked || !onRing { return freePos }
        return CGPoint(x: CGFloat(cos(angle)) * radius, y: CGFloat(sin(angle)) * radius)
    }

    var body: some View {
        ZStack {
            chainBlooms

            if let id = rewards.activeSticker {
                // Confetti trail while she spins it fast.
                ForEach(trail) { bit in
                    Ellipse()
                        .fill(theme.color(bit.tint))
                        .frame(width: 6, height: 9)
                        .rotationEffect(.degrees(bit.spin))
                        .offset(x: CGFloat(cos(bit.angle)) * radius,
                                y: CGFloat(sin(bit.angle)) * radius)
                        .opacity(bit.opacity)
                        .allowsHitTesting(false)
                }

                SparkleBurst(trigger: settleBurst, count: 10)
                    .offset(x: pos.x, y: pos.y)
                    .allowsHitTesting(false)

                // The flower: its 44pt circle IS the hitbox — nothing else here
                // grabs touches, so the ring's own scrub gesture keeps the rest.
                StickerView(id: id, size: 38)
                    .scaleEffect(dragging ? 1.2 : 1)
                    .shadow(color: theme.shadow, radius: dragging || spinning ? 6 : 0)
                    .frame(width: FFSpace.tapMin, height: FFSpace.tapMin)
                    .contentShape(Circle())
                    .offset(x: pos.x, y: pos.y)
                    .highPriorityGesture(stickerDrag)
                    .animation(dragging ? nil : FFMotion.spring, value: dragging)
                    .accessibilityLabel("Your flower sticker")
                    .accessibilityHint("Drag along the ring to spin it, or pull it off and set it anywhere")
            }
        }
        .frame(width: diameter, height: diameter)
        .coordinateSpace(name: "ffRingSticker")
        .overlay(alignment: .top) { teaseBubble }
        .overlay(alignment: .bottom) { bonusBadge }
        .onAppear {
            angle = rewards.stickerAngle
            onRing = rewards.stickerMode != "free"
            // Normalized by RADIUS (not the frame) so a rest keeps the same
            // relation to the band across ring sizes (hero vs preview).
            freePos = CGPoint(x: CGFloat(rewards.stickerX) * radius,
                              y: CGFloat(rewards.stickerY) * radius)
        }
        .sensoryFeedback(.selection, trigger: settleBurst)
    }

    // MARK: the daisy chain — her other worn blooms, riding the same ring

    /// Non-bead worn blooms. Spread: evenly spaced from 12 o'clock. Linked
    /// (chain-together mode): snapped up behind the bead, so a drag or fling
    /// of the bead spins the whole chain as one. Tapping any bloom toggles
    /// the mode, same as the Today tab's chain button.
    @ViewBuilder private var chainBlooms: some View {
        let blooms = rewards.ringChain.filter { $0 != rewards.activeSticker }
        ForEach(Array(blooms.enumerated()), id: \.offset) { i, id in
            let p = chainPos(index: i, count: blooms.count)
            StickerView(id: id, size: 26)
                .frame(width: 36, height: 36)
                .contentShape(Circle())
                .onTapGesture {
                    withAnimation(reduceMotion ? nil : FFMotion.spring) {
                        rewards.chainLinked.toggle()
                    }
                }
                .offset(x: p.x, y: p.y)
                .animation(reduceMotion ? nil : FFMotion.spring, value: rewards.chainLinked)
                .accessibilityLabel("\(RewardsStore.flowers.first { $0.id == id }?.name ?? id) on your ring")
                .accessibilityHint(rewards.chainLinked ? "Spreads your flowers back around the ring"
                                                       : "Snaps your flowers together into a chain")
        }
    }

    /// Hoisted out of the ViewBuilder (release type-check lesson). Linked mode
    /// trails the bead clockwise, one bloom-width apart; with no bead to trail
    /// the blooms fall back to the even spread.
    private func chainPos(index: Int, count: Int) -> CGPoint {
        let a: Double
        if rewards.chainLinked && rewards.activeSticker != nil {
            a = angle + Double(index + 1) * Double(32 / radius)
        } else {
            a = Double(index) / Double(max(count, 1)) * 2 * .pi - .pi / 2
        }
        return CGPoint(x: radius * CGFloat(cos(a)), y: radius * CGFloat(sin(a)))
    }

    private var stickerDrag: some Gesture {
        DragGesture(minimumDistance: 0, coordinateSpace: .named("ffRingSticker"))
            .onChanged { v in
                dragging = true
                spinning = false
                let c = diameter / 2
                let dx = v.location.x - c
                let dy = v.location.y - c
                let r = hypot(dx, dy)
                let onBand = abs(r - radius) <= snapBand
                // A free-resting flower needs real movement before it re-beads —
                // a bare tap must never yank it out of its placed spot.
                let moved = hypot(v.translation.width, v.translation.height) > 8
                if onBand && (onRing || moved) {
                    // On the band: magnetically locked to the ring, concentric.
                    plucked = false
                    onRing = true
                    let a = Double(atan2(dy, dx))
                    if let last = lastDragAngle {
                        var delta = a - last
                        if delta > .pi { delta -= 2 * .pi }
                        if delta < -.pi { delta += 2 * .pi }
                        velocity = delta
                        if !reduceMotion, abs(delta) > 0.06 { emitTrail() }
                    }
                    lastDragAngle = a
                    angle = a
                } else if !onBand && (onRing || plucked || moved) {
                    // Plucked: the flower follows her finger anywhere.
                    plucked = true
                    lastDragAngle = nil
                    velocity = 0
                    let limit = c - 14
                    freePos = CGPoint(x: min(max(dx, -limit), limit),
                                      y: min(max(dy, -limit), limit))
                }
                // else: an untraveled touch on a free-resting flower — hold still.
            }
            .onEnded { _ in
                dragging = false
                lastDragAngle = nil
                if plucked {
                    plucked = false
                    let rest = hypot(freePos.x, freePos.y)
                    if abs(rest - radius) <= snapBand {
                        // Dropped onto the band — it beads back onto the ring.
                        angle = Double(atan2(freePos.y, freePos.x))
                        onRing = true
                        rewards.stickerMode = "ring"
                        settle()
                    } else {
                        onRing = false
                        rewards.stickerX = Double(freePos.x / radius)
                        rewards.stickerY = Double(freePos.y / radius)
                        rewards.stickerMode = "free"
                        settleBurst += 1
                        withAnimation(.easeOut(duration: 0.4)) { trail.removeAll() }
                    }
                } else if onRing {
                    rewards.stickerMode = "ring"
                    if !reduceMotion && abs(velocity) > 0.035 {
                        startSpin()
                    } else {
                        settle()
                    }
                }
                // else: an idle tap on a free-resting flower — leave it be.
            }
    }

    // Posey's gentle nudge after the fifth fling in a row.
    @ViewBuilder private var teaseBubble: some View {
        if showTease {
            HStack(spacing: 8) {
                FlowerMark(size: 22)
                Text("All that spinning, petal. Imagine the petals if we stretched.")
                    .font(ffBody(FFType.xs, weight: .medium))
                    .foregroundStyle(theme.color(.text))
            }
            .padding(.horizontal, 12).padding(.vertical, 8)
            .background(theme.color(.surface), in: Capsule())
            .overlay(Capsule().strokeBorder(theme.color(.line), lineWidth: 1))
            .shadow(color: theme.shadow, radius: 6, y: 2)
            .transition(.move(edge: .top).combined(with: .opacity))
            .allowsHitTesting(false)
        }
    }

    @ViewBuilder private var bonusBadge: some View {
        if showBonus {
            Label("+5, right on your period day", systemImage: "sparkle")
                .font(ffBody(FFType.xs, weight: .bold))
                .foregroundStyle(theme.color(.deep))
                .padding(.horizontal, 12).padding(.vertical, 7)
                .background(theme.color(.phaseMenstrualSoft), in: Capsule())
                .transition(.move(edge: .bottom).combined(with: .opacity))
                .allowsHitTesting(false)
        }
    }

    // The fidget: inertia with gentle friction, confetti while fast.
    private func startSpin() {
        spinning = true
        spins += 1
        if spins >= 5 {
            spins = 0
            Task { @MainActor in
                withAnimation(FFMotion.spring) { showTease = true }
                try? await Task.sleep(for: .seconds(4))
                withAnimation(FFMotion.fast) { showTease = false }
            }
        }
        Task { @MainActor in
            var frames = 0
            while spinning && abs(velocity) > 0.006 && !Task.isCancelled {
                angle += velocity
                velocity *= 0.965
                frames += 1
                if abs(velocity) > 0.05 && frames % 3 == 0 { emitTrail() }
                fadeTrail()
                try? await Task.sleep(for: .milliseconds(16))
            }
            spinning = false
            settle()
        }
    }

    private func settle() {
        velocity = 0
        rewards.stickerAngle = angle
        settleBurst += 1
        checkPeriodLanding()
        Task {
            try? await Task.sleep(for: .seconds(0.8))
            withAnimation(.easeOut(duration: 0.4)) { trail.removeAll() }
        }
    }

    // Landed on the bleed arc? A small once-a-day bonus, with a note.
    private func checkPeriodLanding() {
        guard periodFraction > 0 else { return }
        let deg = angle * 180 / .pi
        var f = ((deg + 90) / 360).truncatingRemainder(dividingBy: 1)
        if f < 0 { f += 1 }
        guard f < periodFraction else { return }
        let df = DateFormatter(); df.dateFormat = "yyyy-MM-dd"
        guard rewards.awardPeriodLanding(dateKey: df.string(from: Date())) else { return }
        Task { @MainActor in
            withAnimation(FFMotion.spring) { showBonus = true }
            try? await Task.sleep(for: .seconds(3))
            withAnimation(FFMotion.fast) { showBonus = false }
        }
    }

    private func emitTrail() {
        let i = trail.count
        trail.append(TrailBit(
            angle: angle - Double(velocity) * 2,
            spin: Double((i &* 97) % 70) - 35,
            tint: i % 3 == 0 ? .flowerCenter : (i % 2 == 0 ? .primary : .phaseFollicular)
        ))
        if trail.count > 16 { trail.removeFirst(trail.count - 16) }
    }

    private func fadeTrail() {
        for idx in trail.indices { trail[idx].opacity *= 0.9 }
        trail.removeAll { $0.opacity < 0.05 }
    }
}

private struct TrailBit: Identifiable {
    let id = UUID()
    var angle: Double
    var spin: Double
    var tint: Tok
    var opacity: Double = 0.85
}

// RingChainView — her daisy chain: every bloom she's chained in the shop rides
// the ring together, evenly spaced from 12 o'clock. Purely decorative (the one
// ACTIVE sticker stays the draggable bead); no touches stolen from the scrub.
struct RingChainView: View {
    let chain: [String]
    let radius: CGFloat

    var body: some View {
        ZStack {
            ForEach(Array(chain.enumerated()), id: \.offset) { i, id in
                let p = position(for: i)
                StickerView(id: id, size: 19)
                    .offset(x: p.x, y: p.y)
            }
        }
        .allowsHitTesting(false)
        .accessibilityHidden(true)
    }

    // Hoisted out of the ViewBuilder: the inline CGFloat/Double trig mix made
    // the release build's type-checker time out (simulator builds squeaked by).
    private func position(for index: Int) -> CGPoint {
        let angle: Double = Double(index) / Double(max(chain.count, 1)) * 2 * .pi - .pi / 2
        return CGPoint(x: radius * CGFloat(cos(angle)),
                       y: radius * CGFloat(sin(angle)))
    }
}

File: web/components/ringSticker.js
// RingSticker — port of App/Components/RingSticker.swift. Her flower rides the
// cycle ring like a bead on a bangle. The hitbox is the flower itself (the rest
// of the ring keeps its own tap). Dragging along the ring band keeps it
// magnetically locked to the ring — always concentric, never offset; fling it
// and it keeps spinning like a fidget spinner, shedding confetti while it's
// fast. Pull it clearly off the band and it PLUCKS free: drop it inside the ring
// or off to the side and it rests there (persisted); carry it back to the band
// and it snaps on again. Reduce Motion: drag still works, no inertia, no trail.

import { rewards, FLOWERS } from "../core/rewards.js";
import { keyFromDate } from "../core/dates.js";

const React = window.React;
const { useState, useRef, useEffect } = React;
const html = window.htm.bind(React.createElement);
const UI = window.FlowtierDesignSystem_6b3d0d || {};

/// Radius of the drawn track for a ring of `size` — the DS CycleRing draws its
/// arcs at r=84 in a 200 viewBox, so the flower beads onto the visible ring,
/// concentric with it. Mirrors CycleRing.trackRadius(for:) in Swift.
export const trackRadius = (size) => (84 * size) / 200;

const SNAP = 26;       // how far off the ring still counts as "on the band"
const FRAME_PAD = 96;  // Swift: diameter = radius * 2 + 96 — room for a free rest
const TAP = 44;        // FFSpace.tapMin — the flower's hitbox
const MAX_TRAIL = 16;

const reduceMotion = () =>
  typeof window.matchMedia === "function" &&
  window.matchMedia("(prefers-reduced-motion: reduce)").matches;

// The worn sticker, drawn inline (no emoji, no external asset) — the same
// procedural bloom the garden shop uses, so a flower reads the same in both.
function Bloom({ id, size }) {
  if (id === "posey" && UI.FlowerMark) return html`<${UI.FlowerMark} size=${size} />`;
  const idx = Math.max(FLOWERS.findIndex((f) => f.id === id), 0);
  const petals = 5 + (idx % 6);
  const fills = ["var(--primary)", "var(--primary-strong)", "var(--flower-center)", "var(--good)"];
  const fill = fills[idx % fills.length];
  const c = size / 2;
  const ring = size * 0.26;
  const nodes = [];
  for (let i = 0; i < petals; i++) {
    const a = (i / petals) * Math.PI * 2;
    const px = c + Math.cos(a) * ring;
    const py = c + Math.sin(a) * ring;
    nodes.push(html`<ellipse key=${i} cx=${px} cy=${py} rx=${size * 0.15} ry=${size * 0.26}
      transform=${`rotate(${(a * 180) / Math.PI + 90} ${px} ${py})`} fill=${fill} />`);
  }
  return html`<svg width=${size} height=${size} viewBox=${`0 0 ${size} ${size}`} aria-hidden="true"
      style=${{ display: "block" }}>
    ${nodes}
    <circle cx=${c} cy=${c} r=${size * 0.17} fill="var(--flower-center)" />
  </svg>`;
}

// RingChainView (RingSticker.swift) — her daisy chain: every bloom she's
// chained in the shop rides the ring together, evenly spaced from 12 o'clock.
// Purely decorative (the one ACTIVE sticker stays the draggable bead); no
// touches stolen from the scrub. Subscribes to rewards itself so chaining in
// the garden shop shows up the moment she's back on Today.
export function RingChain({ radius }) {
  const [, force] = useState(0);
  useEffect(() => rewards.subscribe(() => force((n) => n + 1)), []);
  const chain = rewards.ringChain ?? [];
  if (!chain.length) return null;
  return html`
    <div aria-hidden="true" style=${{
      position: "absolute", left: "50%", top: "50%", width: 0, height: 0, pointerEvents: "none",
    }}>
      ${chain.map((id, i) => {
        const a = (i / Math.max(chain.length, 1)) * 2 * Math.PI - Math.PI / 2;
        return html`<div key=${i} style=${{
          position: "absolute",
          transform: `translate(-50%, -50%) translate(${radius * Math.cos(a)}px, ${radius * Math.sin(a)}px)`,
        }}><${Bloom} id=${id} size=${19} /></div>`;
      })}
    </div>`;
}

export function RingSticker({ radius, periodFraction = 0 }) {
  const [, force] = useState(0);
  useEffect(() => rewards.subscribe(() => force((n) => n + 1)), []);

  // Her last resting place, from the shared garden blob (the same keys the iOS
  // build reads). The free rest is normalized by RADIUS, not the frame, so it
  // keeps the same relation to the band across ring sizes (hero vs preview).
  const [seed] = useState(() => ({
    angle: rewards.stickerAngle,
    free: { x: rewards.stickerX * radius, y: rewards.stickerY * radius },
    onRing: rewards.stickerMode !== "free",
  }));
  const [angle, setAngle] = useState(seed.angle);
  const [free, setFree] = useState(seed.free);
  const [onRing, setOnRing] = useState(seed.onRing);
  const [plucked, setPlucked] = useState(false);
  const [dragging, setDragging] = useState(false);
  const [trail, setTrail] = useState([]);
  const [pop, setPop] = useState(0);
  const [tease, setTease] = useState(false);
  const [bonus, setBonus] = useState(false);

  // React may batch a pointermove's state update past the pointerup that follows
  // it (a fast flick lands both in one frame), so every value the gesture READS
  // lives in a ref; the matching state is only there to paint.
  const wrap = useRef(null);
  const angRef = useRef(seed.angle);
  const freeRef = useRef(seed.free);
  const ringRef = useRef(seed.onRing);
  const vel = useRef(0);
  const lastAng = useRef(null);
  const drag = useRef({ active: false, x: 0, y: 0, plucked: false });
  const spins = useRef(0);
  const raf = useRef(0);
  const timers = useRef([]);

  const putAngle = (a) => { angRef.current = a; setAngle(a); };
  const putFree = (x, y) => { freeRef.current = { x, y }; setFree({ x, y }); };
  const putRing = (v) => { ringRef.current = v; setOnRing(v); };
  const putPlucked = (v) => { drag.current.plucked = v; setPlucked(v); };

  const after = (ms, fn) => timers.current.push(setTimeout(fn, ms));
  useEffect(() => () => {
    cancelAnimationFrame(raf.current);
    timers.current.forEach(clearTimeout);
  }, []);

  const id = rewards.activeSticker;
  const diameter = radius * 2 + FRAME_PAD;
  const pos = plucked || !onRing ? free : { x: Math.cos(angle) * radius, y: Math.sin(angle) * radius };

  const emitTrail = () => setTrail((t) => {
    const i = t.length;
    const next = t.concat({
      key: Date.now() + i,
      angle: angRef.current - vel.current * 2,
      spin: ((i * 97) % 70) - 35,
      tint: i % 3 === 0 ? "var(--flower-center)" : i % 2 === 0 ? "var(--primary)" : "var(--phase-follicular)",
      opacity: 0.85,
    });
    return next.slice(-MAX_TRAIL);
  });

  const fadeTrail = () => setTrail((t) =>
    t.map((b) => ({ ...b, opacity: b.opacity * 0.9 })).filter((b) => b.opacity >= 0.05));

  // Landed on the bleed arc? A small once-a-day bonus, with a note. The arc runs
  // from 12 o'clock clockwise, so the angle is turned a quarter before testing.
  const checkPeriodLanding = (a) => {
    if (!(periodFraction > 0)) return;
    const f = (((((a * 180) / Math.PI + 90) / 360) % 1) + 1) % 1;
    if (f >= periodFraction) return;
    if (!rewards.awardPeriodLanding(keyFromDate(new Date()))) return;
    setBonus(true);
    after(3000, () => setBonus(false));
  };

  const settle = (a) => {
    vel.current = 0;
    rewards.setStickerAngle(a);
    setPop((n) => n + 1);
    checkPeriodLanding(a);
    after(800, () => setTrail([]));
  };

  // The fidget: inertia with gentle friction, confetti while fast.
  const startSpin = () => {
    spins.current += 1;
    if (spins.current >= 5) {
      spins.current = 0;
      setTease(true);
      after(4000, () => setTease(false));
    }
    let frames = 0;
    const step = () => {
      putAngle(angRef.current + vel.current);
      vel.current *= 0.965;
      frames += 1;
      if (Math.abs(vel.current) > 0.05 && frames % 3 === 0) emitTrail();
      fadeTrail();
      if (Math.abs(vel.current) > 0.006) raf.current = requestAnimationFrame(step);
      else settle(angRef.current);
    };
    raf.current = requestAnimationFrame(step);
  };

  const center = () => {
    const b = wrap.current.getBoundingClientRect();
    return { x: b.left + b.width / 2, y: b.top + b.height / 2 };
  };

  const onDown = (e) => {
    e.stopPropagation();
    e.currentTarget.setPointerCapture(e.pointerId);
    cancelAnimationFrame(raf.current);
    drag.current = { active: true, x: e.clientX, y: e.clientY };
    setDragging(true);
  };

  const onMove = (e) => {
    if (!drag.current.active) return;
    const c = center();
    const dx = e.clientX - c.x;
    const dy = e.clientY - c.y;
    const onBand = Math.abs(Math.hypot(dx, dy) - radius) <= SNAP;
    // A free-resting flower needs real movement before it re-beads — a bare tap
    // must never yank it out of its placed spot.
    const moved = Math.hypot(e.clientX - drag.current.x, e.clientY - drag.current.y) > 8;
    if (onBand && (ringRef.current || moved)) {
      // On the band: magnetically locked to the ring, concentric.
      putPlucked(false);
      putRing(true);
      const a = Math.atan2(dy, dx);
      if (lastAng.current != null) {
        let delta = a - lastAng.current;
        if (delta > Math.PI) delta -= 2 * Math.PI;
        if (delta < -Math.PI) delta += 2 * Math.PI;
        vel.current = delta;
        if (!reduceMotion() && Math.abs(delta) > 0.06) emitTrail();
      }
      lastAng.current = a;
      putAngle(a);
    } else if (!onBand && (ringRef.current || drag.current.plucked || moved)) {
      // Plucked: the flower follows her finger anywhere.
      putPlucked(true);
      lastAng.current = null;
      vel.current = 0;
      const limit = diameter / 2 - 14;
      putFree(Math.min(Math.max(dx, -limit), limit), Math.min(Math.max(dy, -limit), limit));
    }
    // else: an untraveled touch on a free-resting flower — hold still.
  };

  const onUp = () => {
    if (!drag.current.active) return;
    drag.current.active = false;
    setDragging(false);
    lastAng.current = null;
    const rest = freeRef.current;
    if (drag.current.plucked) {
      putPlucked(false);
      if (Math.abs(Math.hypot(rest.x, rest.y) - radius) <= SNAP) {
        // Dropped onto the band — it beads back onto the ring.
        const a = Math.atan2(rest.y, rest.x);
        putAngle(a);
        putRing(true);
        rewards.setStickerMode("ring");
        settle(a);
      } else {
        putRing(false);
        rewards.setStickerPosition(rest.x / radius, rest.y / radius);
        rewards.setStickerMode("free");
        setPop((n) => n + 1);
        setTrail([]);
      }
    } else if (ringRef.current) {
      rewards.setStickerMode("ring");
      if (!reduceMotion() && Math.abs(vel.current) > 0.035) startSpin();
      else settle(angRef.current);
    }
    // else: an idle tap on a free-resting flower — leave it be.
  };

  if (!id) return null;

  const at = (x, y) => `translate(-50%, -50%) translate(${x}px, ${y}px)`;

  return html`
    <div ref=${wrap} style=${{
      position: "absolute", left: "50%", top: "50%",
      transform: "translate(-50%, -50%)",
      width: diameter, height: diameter, pointerEvents: "none",
    }}>
      ${trail.map((b) => html`
        <span key=${b.key} style=${{
          position: "absolute", left: "50%", top: "50%", width: 6, height: 9,
          borderRadius: "50%", background: b.tint, opacity: b.opacity,
          transform: `${at(Math.cos(b.angle) * radius, Math.sin(b.angle) * radius)} rotate(${b.spin}deg)`,
        }} />`)}

      <div key=${`pop-${pop}`} onPointerDown=${onDown} onPointerMove=${onMove}
        onPointerUp=${onUp} onPointerCancel=${onUp}
        role="button" tabindex="0"
        aria-label="Your flower sticker"
        aria-description="Drag along the ring to spin it, or pull it off and set it anywhere"
        style=${{
          position: "absolute", left: "50%", top: "50%",
          width: TAP, height: TAP, borderRadius: "50%",
          display: "flex", alignItems: "center", justifyContent: "center",
          transform: `${at(pos.x, pos.y)} scale(${dragging ? 1.2 : 1})`,
          transition: dragging ? "none" : "transform var(--dur-base) var(--ease-signature)",
          filter: dragging ? "drop-shadow(0 2px 6px var(--shadow))" : "none",
          animation: pop && !dragging ? "flowtier-pop var(--dur-base) var(--ease-signature)" : "none",
          pointerEvents: "auto", touchAction: "none", cursor: "grab",
        }}>
        <${Bloom} id=${id} size=${30} />
      </div>

      ${tease && html`
        <div style=${{
          position: "absolute", top: 0, left: "50%", transform: "translateX(-50%)",
          display: "flex", alignItems: "center", gap: 8, whiteSpace: "nowrap",
          padding: "8px 12px", borderRadius: 999, background: "var(--surface)",
          border: "1px solid var(--line)", boxShadow: "0 2px 6px var(--shadow)",
          fontSize: "var(--text-xs)", fontWeight: 500, color: "var(--text)",
        }}>
          ${UI.FlowerMark && html`<${UI.FlowerMark} size=${22} />`}
          All that spinning, petal — imagine the petals if we stretched.
        </div>`}

      ${bonus && html`
        <div style=${{
          position: "absolute", bottom: 0, left: "50%", transform: "translateX(-50%)",
          whiteSpace: "nowrap", padding: "7px 12px", borderRadius: 999,
          background: "var(--phase-menstrual-soft)", color: "var(--deep)",
          fontSize: "var(--text-xs)", fontWeight: 700,
        }}>+5 — right on your period day</div>`}
    </div>`;
}

export default RingSticker;
